src/platform/K32W/KeyValueStoreManagerImpl.cpp

```diff
@@ -24,76 +24,188 @@
+        key_id = it->second;
 
-    err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(
-        chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size, read_bytes);
+        err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size, read_bytes);
+        *read_bytes_size = read_bytes;
 
-    *read_bytes_size = read_bytes;
+        ChipLogProgress(DeviceLayer, "KVS, get key [%s] with PDM key: %i", key, key_id);
+    }
 
 exit:
     return err;
 }
 
 CHIP_ERROR KeyValueStoreManagerImpl::_Put(const char * key, const void * value, size_t value_size)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    std::hash<std::string> hash_fn;
-    uint16_t key_id;
+    CHIP_ERROR err = CHIP_ERROR_INVALID_ARGUMENT;
+    uint8_t key_id;
+    bool_t put_key     = FALSE;
     uint8_t pdm_id_kvs = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
-
-    VerifyOrExit((key != NULL) && (value != NULL), err = CHIP_ERROR_INVALID_ARGUMENT);
-    key_id = hash_fn(key) % MAX_NO_OF_KEYS;
-
-    ChipLogProgress(DeviceLayer, "KVS, put key id:: %i", key_id);
-
-    err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueBin(
-        chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size);
+    std::unordered_map<std::string, uint8_t>::const_iterator it;
+
+    VerifyOrExit((key != NULL) && (value != NULL) && (RestoreFromFlash() == CHIP_NO_ERROR), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    if ((it = g_kvs_map.find(key)) == g_kvs_map.end()) /* new key */
+    {
+        for (key_id = 0; key_id < MAX_NO_OF_KEYS; key_id++)
+        {
+            std::list<uint8_t>::iterator iter = std::find(g_key_ids_list.begin(), g_key_ids_list.end(), key_id);
+
+            if (iter == g_key_ids_list.end())
+            {
+                g_key_ids_list.push_back(key_id);
+
+                put_key = TRUE;
+                break;
+            }
+        }
+    }
+    else /* overwrite key */
+    {
+        put_key = TRUE;
+        key_id  = it->second;
+    }
+
+    if (put_key)
+    {
+        ChipLogProgress(DeviceLayer, "KVS, put key [%s] with PDM key: %i", key, key_id);
+
+        g_kvs_map.insert(std::make_pair(std::string(key), key_id));
+        err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueBin(
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size);
+
+        /* save the 'key' in flash such that it can be retrieved later on */
+        if (err == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, save in flash key [%s] with PDM key: %i", key, key_id + MAX_NO_OF_KEYS);
+            err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueStr(
+                chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + MAX_NO_OF_KEYS), key, strlen(key));
+        }
+    }
```

Comment:
What if false? it sounds like we want to return some error. In this case single exit point makes  the code hard to follow. I see it probably returns CHIP_ERROR_INVALID_ARGUMENT ... but why is that correct?

Reply:
With the newly added assert (assert(g_key_ids_set.size() < kMaxNumberOfKeys)), the condition -> if (put_key) should always be true.

Reply:
The assert can:

* get us into a reset loop if existing elements in the persistent store trigger storage of additional elements
* be compiled away and ignored in production code

So this is probably not the greatest solution.

I would think that the right thing to do here is to return CHIP_ERROR_NO_MEMORY, not assert.  It's easy enough to do.

Reply:
Added suggestion here:

https://github.com/project-chip/connectedhomeip/pull/9489/files#r744129165